File: 03-prototype/js/drug-engine.js
/** @typedef {'green'|'yellow'|'red'} SafetyLevel */

/**
 * @param {object} db
 * @param {string} raw
 */
export function normalizeText(raw) {
  return raw
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .trim();
}

/**
 * @param {object} db
 * @param {string} condition
 */
export function detectUrgent(db, condition) {
  const n = normalizeText(condition);
  return db.urgentKeywords.some((kw) => n.includes(normalizeText(kw)));
}

/**
 * @param {string} a
 * @param {string} b
 */
function levenshteinRatio(a, b) {
  if (!a || !b) return 0;
  if (a === b) return 1;
  const m = a.length;
  const n = b.length;
  const dp = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(0));
  for (let i = 0; i <= m; i++) dp[i][0] = i;
  for (let j = 0; j <= n; j++) dp[0][j] = j;
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      dp[i][j] = Math.min(dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + cost);
    }
  }
  return 1 - dp[m][n] / Math.max(m, n);
}

function labelMatchesQuery(label, query) {
  const n = normalizeText(label);
  const q = normalizeText(query);
  if (!q) return false;
  if (n === q) return true;
  if (levenshteinRatio(q, n) >= 0.78) return true;
  if (q.length >= 3 && n.startsWith(q)) return true;
  if (q.length >= 4 && n.includes(q)) return true;
  return false;
}

/**
 * @param {object} db
 * @param {string} query
 */
export function findDrugCandidates(db, query) {
  const q = normalizeText(query);
  if (!q) return [];

  const matches = db.drugs.filter((drug) => {
    const labels = [drug.name, drug.activeIngredient, ...(drug.aliases || [])];
    return labels.some((label) => labelMatchesQuery(label, q));
  });

  const unique = [];
  const seen = new Set();
  for (const d of matches) {
    if (!seen.has(d.id)) {
      seen.add(d.id);
      unique.push(d);
    }
  }
  return unique;
}

/** Khớp chính xác tên/hoạt chất/alias — không fuzzy */
export function isExactDrugInDb(db, query) {
  const q = normalizeText(query);
  if (!q) return false;
  return db.drugs.some(
    (d) =>
      normalizeText(d.name) === q ||
      normalizeText(d.activeIngredient) === q ||
      (d.aliases || []).some((a) => normalizeText(a) === q)
  );
}

/**
 * @param {object} db
 * @param {string} query
 * @param {string|null} drugId
 */
export function resolveExactDrug(db, query, drugId = null) {
  if (drugId) {
    return db.drugs.find((d) => d.id === drugId) || null;
  }
  if (!isExactDrugInDb(db, query)) return null;
  const candidates = findDrugCandidates(db, query).filter(
    (d) =>
      normalizeText(d.name) === normalizeText(query) ||
      normalizeText(d.activeIngredient) === normalizeText(query) ||
      (d.aliases || []).some((a) => normalizeText(a) === normalizeText(query))
  );
  return candidates.length >= 1 ? candidates[0] : null;
}

/**
 * @param {object} drug
 * @param {string} condition
 * @param {{ age?: number|null, gender?: 'male'|'female'|null }} [patient]
 */
export function evaluateSafety(drug, condition, patient = {}) {
  const n = normalizeText(condition);
  const { age = null, gender = null } = patient;
  /** @type {{ level: SafetyLevel, note: string }[]} */
  const hits = [];

  for (const rule of drug.conditionRules || []) {
    const matched = rule.match.some((m) => n.includes(normalizeText(m)));
    if (matched) hits.push({ level: rule.level, note: rule.note });
  }

  for (const ci of drug.contraindications || []) {
    const parts = normalizeText(ci).split(/\s+/);
    if (parts.some((p) => p.length > 3 && n.includes(p))) {
      hits.push({
        level: 'red',
        note: `Liên quan chống chỉ định: ${ci}`,
      });
    }
  }

  if (age != null && Number.isFinite(age)) {
    for (const rule of drug.ageRules || []) {
      if (rule.minAge != null && age < rule.minAge) continue;
      if (rule.maxAge != null && age > rule.maxAge) continue;
      hits.push({ level: rule.level, note: rule.note });
    }
  }

  if (gender && age != null) {
    for (const rule of drug.genderRules || []) {
      if (rule.gender !== gender) continue;
      if (rule.minAge != null && age < rule.minAge) continue;
      if (rule.maxAge != null && age > rule.maxAge) continue;
      hits.push({ level: rule.level, note: rule.note });
    }
  }

  if (hits.length === 0) {
    return {
      level: /** @type {SafetyLevel} */ ('yellow'),
      summary:
        'Chưa đủ dữ liệu demo để kết luận an toàn — nên hỏi dược sĩ Long Châu.',
      matchedRules: [],
    };
  }

  const order = { red: 3, yellow: 2, green: 1 };
  hits.sort((a, b) => order[b.level] - order[a.level]);
  const top = hits[0];

  const levelLabels = {
    green: 'An toàn có điều kiện',
    yellow: 'Thận trọng — nên hỏi dược sĩ',
    red: 'Không nên tự ý — cần chuyên môn / cấp cứu',
  };

  return {
    level: top.level,
    summary: levelLabels[top.level],
    matchedRules: hits,
  };
}

/** @param {unknown} raw @returns {number|null} */
export function parseAge(raw) {
  const m = String(raw ?? '').match(/\d+/);
  if (!m) return null;
  const age = Number(m[0]);
  if (!Number.isFinite(age) || age < 0 || age > 120) return null;
  return age;
}

/** @param {unknown} raw @returns {'male'|'female'|null} */
export function parseGender(raw) {
  const t = normalizeText(String(raw ?? ''));
  if (['nam', 'male', 'boy', 'dan ong'].includes(t) || t === '__gender_male__') return 'male';
  if (['nu', 'nữ', 'female', 'girl', 'dan ba'].includes(t) || t === '__gender_female__') return 'female';
  return null;
}

/**
 * Trích tuổi + giới tính từ câu chat (vd: "35 tuổi, nữ", "35 tuổi nữ")
 * @param {string} text
 * @returns {{ age: number|null, gender: 'male'|'female'|null }}
 */
export function extractPatientFromMessage(text) {
  const t = String(text || '').trim();
  let age = null;
  let gender = null;

  const ageMatch =
    t.match(/(\d{1,3})\s*(?:tuổi|tuoi)/i) ||
    t.match(/(\d{1,3})\s*t(?=[\s,;·]|$)/i) ||
    t.match(/^(\d{1,3})(?:\s*[,;·]|\s+(?:nữ|nu|nam)\b)/i);
  if (ageMatch) age = parseAge(ageMatch[1]);

  const chunks = [t, ...t.split(/[,;·]/).map((s) => s.trim()).filter(Boolean)];
  for (const chunk of chunks) {
    const g = parseGender(chunk);
    if (g) gender = g;
  }

  const n = normalizeText(t);
  if (!gender) {
    if (/(^|[\s,;])nu($|[\s,;])|nữ|female|phu nu|dan ba/.test(n)) gender = 'female';
    else if (/(^|[\s,;])nam($|[\s,;])|male|dan ong/.test(n) && !/nam dinh/.test(n)) gender = 'male';
  }

  if (!age) {
    const bare = t.match(/^(\d{1,3})$/);
    if (bare) age = parseAge(bare[1]);
  }

  return { age, gender };
}

export function formatGender(gender) {
  return gender === 'male' ? 'Nam' : gender === 'female' ? 'Nữ' : '—';
}

function buildAgeAppropriateSummary(evaluation, age, gender) {
  const ageGenderHits = (evaluation.matchedRules || []).filter((r) =>
    /tuổi|trẻ|trẻ em|cao tuổi|mang thai|nữ|nam/i.test(r.note)
  );
  if (ageGenderHits.length) return ageGenderHits.map((h) => h.note).join(' ');
  if (age != null && gender) {
    return `Đã đối chiếu ${age} tuổi, giới tính ${formatGender(gender)} — chưa có rule độ tuổi cụ thể trong DB demo.`;
  }
  return null;
}

/**
 * @param {object} db
 */
export async function loadDrugDatabase() {
  const res = await fetch('./data/drugs-demo.json');
  if (!res.ok) throw new Error('Không tải được drugs-demo.json — chạy qua local server.');
  return res.json();
}

export function buildSafetyCardPayload(drug, condition, evaluation, userName, patient = {}) {
  const { age = null, gender = null } = patient;
  return {
    userName: userName || null,
    condition,
    age,
    gender,
    genderLabel: formatGender(gender),
    drugName: drug.name,
    activeIngredient: drug.activeIngredient,
    indications: drug.indications,
    contraindications: drug.contraindications,
    warnings: drug.warnings,
    safetyLevel: evaluation.level,
    safetySummary: evaluation.summary,
    ageAppropriate: buildAgeAppropriateSummary(evaluation, age, gender),
    matchedRules: evaluation.matchedRules,
    source: drug.source,
    timestamp: new Date().toISOString(),
  };
}

File: 03-prototype/python/drug_engine.py
"""Drug safety engine — Python port for Streamlit."""

from __future__ import annotations

import json
import re
import unicodedata
from difflib import SequenceMatcher
from pathlib import Path
from typing import Any, Literal

SafetyLevel = Literal["green", "yellow", "red"]
Gender = Literal["male", "female"]

_DB: dict | None = None
_DB_PATH = Path(__file__).resolve().parent.parent / "data" / "drugs-demo.json"


def normalize_text(raw: str) -> str:
    t = unicodedata.normalize("NFD", str(raw or "").lower())
    t = "".join(c for c in t if unicodedata.category(c) != "Mn")
    return t.strip()


def load_db() -> dict:
    global _DB
    if _DB is None:
        with open(_DB_PATH, encoding="utf-8") as f:
            _DB = json.load(f)
    return _DB


def parse_age(raw: Any) -> int | None:
    m = re.search(r"\d+", str(raw or ""))
    if not m:
        return None
    age = int(m.group())
    return age if 0 <= age <= 120 else None


def parse_gender(raw: Any) -> Gender | None:
    t = normalize_text(str(raw or ""))
    if t in {"nam", "male", "boy", "dan ong"}:
        return "male"
    if t in {"nu", "nữ", "female", "girl", "dan ba"}:
        return "female"
    return None


def extract_patient(text: str) -> tuple[int | None, Gender | None]:
    t = text.strip()
    age = None
    gender = None

    age_match = (
        re.search(r"(\d{1,3})\s*(?:tuổi|tuoi)", t, re.I)
        or re.search(r"(\d{1,3})\s*t(?=[\s,;·]|$)", t, re.I)
        or re.search(r"^(\d{1,3})(?:\s*[,;·]|\s+(?:nữ|nu|nam)\b)", t, re.I)
    )
    if age_match:
        age = parse_age(age_match.group(1))

    for chunk in [t, *[p.strip() for p in re.split(r"[,;·]", t) if p.strip()]]:
        g = parse_gender(chunk)
        if g:
            gender = g

    n = normalize_text(t)
    if not gender:
        if re.search(r"(^|[\s,;])nu($|[\s,;])|nữ|female|phu nu|dan ba", n):
            gender = "female"
        elif re.search(r"(^|[\s,;])nam($|[\s,;])|male|dan ong", n) and "nam dinh" not in n:
            gender = "male"

    return age, gender


def _ratio(a: str, b: str) -> float:
    if not a or not b:
        return 0.0
    return SequenceMatcher(None, a, b).ratio()


def _label_matches(label: str, query: str) -> bool:
    n, q = normalize_text(label), normalize_text(query)
    if not q:
        return False
    if n == q:
        return True
    if _ratio(q, n) >= 0.78:
        return True
    if len(q) >= 3 and n.startswith(q):
        return True
    if len(q) >= 4 and q in n:
        return True
    return False


def find_drug_candidates(db: dict, query: str) -> list[dict]:
    q = normalize_text(query)
    if not q:
        return []
    seen: set[str] = set()
    out: list[dict] = []
    for drug in db["drugs"]:
        labels = [drug["name"], drug["activeIngredient"], *(drug.get("aliases") or [])]
        if any(_label_matches(lbl, q) for lbl in labels):
            if drug["id"] not in seen:
                seen.add(drug["id"])
                out.append(drug)
    return out


def is_exact_drug(db: dict, query: str) -> bool:
    q = normalize_text(query)
    if not q:
        return False
    for d in db["drugs"]:
        if normalize_text(d["name"]) == q or normalize_text(d["activeIngredient"]) == q:
            return True
        if any(normalize_text(a) == q for a in d.get("aliases") or []):
            return True
    return False


def resolve_exact_drug(db: dict, query: str, drug_id: str | None = None) -> dict | None:
    if drug_id:
        return next((d for d in db["drugs"] if d["id"] == drug_id), None)
    if not is_exact_drug(db, query):
        return None
    cands = [
        d
        for d in find_drug_candidates(db, query)
        if normalize_text(d["name"]) == normalize_text(query)
        or normalize_text(d["activeIngredient"]) == normalize_text(query)
        or any(normalize_text(a) == normalize_text(query) for a in d.get("aliases") or [])
    ]
    return cands[0] if cands else None


def detect_urgent(db: dict, condition: str) -> bool:
    n = normalize_text(condition)
    return any(normalize_text(kw) in n for kw in db.get("urgentKeywords") or [])


def evaluate_safety(
    drug: dict, condition: str, patient: dict | None = None
) -> dict[str, Any]:
    patient = patient or {}
    age = patient.get("age")
    gender = patient.get("gender")
    n = normalize_text(condition)
    hits: list[dict] = []

    for rule in drug.get("conditionRules") or []:
        if any(normalize_text(m) in n for m in rule.get("match") or []):
            hits.append({"level": rule["level"], "note": rule["note"]})

    for ci in drug.get("contraindications") or []:
        parts = normalize_text(ci).split()
        if any(len(p) > 3 and p in n for p in parts):
            hits.append({"level": "red", "note": f"Liên quan chống chỉ định: {ci}"})

    if age is not None:
        for rule in drug.get("ageRules") or []:
            if rule.get("minAge") is not None and age < rule["minAge"]:
                continue
            if rule.get("maxAge") is not None and age > rule["maxAge"]:
                continue
            hits.append({"level": rule["level"], "note": rule["note"]})

    if gender and age is not None:
        for rule in drug.get("genderRules") or []:
            if rule.get("gender") != gender:
                continue
            if rule.get("minAge") is not None and age < rule["minAge"]:
                continue
            if rule.get("maxAge") is not None and age > rule["maxAge"]:
                continue
            hits.append({"level": rule["level"], "note": rule["note"]})

    if not hits:
        return {
            "level": "yellow",
            "summary": "Chưa đủ dữ liệu demo — nên hỏi dược sĩ Long Châu.",
            "matchedRules": [],
        }

    order = {"red": 3, "yellow": 2, "green": 1}
    hits.sort(key=lambda h: order.get(h["level"], 0), reverse=True)
    top = hits[0]
    labels = {
        "green": "An toàn có điều kiện",
        "yellow": "Thận trọng — nên hỏi dược sĩ",
        "red": "Không nên tự ý — cần chuyên môn / cấp cứu",
    }
    return {"level": top["level"], "summary": labels.get(top["level"], labels["yellow"]), "matchedRules": hits}


def build_safety_card(drug: dict, condition: str, evaluation: dict, patient: dict | None = None) -> dict:
    patient = patient or {}
    return {
        "drugName": drug["name"],
        "activeIngredient": drug["activeIngredient"],
        "condition": condition,
        "age": patient.get("age"),
        "gender": patient.get("gender"),
        "indications": drug.get("indications", ""),
        "contraindications": drug.get("contraindications") or [],
        "warnings": drug.get("warnings") or [],
        "safetyLevel": evaluation["level"],
        "safetySummary": evaluation["summary"],
        "matchedRules": evaluation.get("matchedRules") or [],
        "source": drug.get("source", "Demo DB"),
        "sources": [{"title": drug.get("source", "Demo DB"), "link": None, "snippet": "DB nội bộ Long Châu (demo)"}],
    }


def suggest_drug_names(db: dict, query: str, limit: int = 5) -> list[dict]:
    q = query.strip()
    if not q:
        return []
    seen: set[str] = set()
    out: list[dict] = []
    for drug in find_drug_candidates(db, q):
        if drug["id"] in seen:
            continue
        seen.add(drug["id"])
        out.append({"drug": drug, "name": drug["name"], "reason": "Khớp / gần giống trong DB"})
        if len(out) >= limit:
            break
    return out


def suggest_drugs_for_condition(db: dict, condition: str) -> list[dict]:
    n = normalize_text(condition)
    hits: list[dict] = []
    for drug in db["drugs"]:
        for rule in drug.get("conditionRules") or []:
            if any(normalize_text(m) in n for m in rule.get("match") or []):
                score = {"green": 3, "yellow": 2, "red": 1}.get(rule["level"], 1)
                hits.append({"drug": drug, "level": rule["level"], "note": rule["note"], "score": score})
                break
    hits.sort(key=lambda h: h["score"], reverse=True)
    seen: set[str] = set()
    out: list[dict] = []
    for h in hits:
        if h["drug"]["id"] in seen:
            continue
        seen.add(h["drug"]["id"])
        out.append(h)
    return out[:5]


def lookup_local(drug: dict, condition: str, patient: dict) -> dict:
    ev = evaluate_safety(drug, condition, patient)
    card = build_safety_card(drug, condition, ev, patient)
    return {"status": "ok", "mode": "database", "card": card, "drug": drug, "normalizedQuery": drug["name"]}


def build_structured(card: dict, drug: dict | None = None) -> dict:
    warnings = card.get("warnings") or []
    ci = card.get("contraindications") or []
    dosage = (drug or {}).get("dosage") or "Theo nhãn thuốc hoặc chỉ định dược sĩ/bác sĩ."
    return {
        "drugName": card["drugName"],
        "activeIngredient": card["activeIngredient"],
        "safetyLevel": card["safetyLevel"],
        "safetySummary": card["safetySummary"],
        "condition": card.get("condition", ""),
        "sections": [
            {"n": 1, "title": "Tên thuốc / hoạt chất", "body": f"{card['drugName']} ({card['activeIngredient']})"},
            {"n": 2, "title": "Công dụng", "body": card.get("indications") or "—"},
            {"n": 3, "title": "Cách dùng / liều tham khảo", "body": dosage},
            {"n": 4, "title": "Tác dụng phụ", "body": "; ".join(warnings) if warnings else "Theo tờ hướng dẫn."},
            {"n": 5, "title": "Chống chỉ định / tương tác", "body": "; ".join(ci) if ci else "Hỏi dược sĩ nếu có bệnh nền."},
        ],
        "matchedRules": card.get("matchedRules") or [],
        "sources": card.get("sources") or [],
    }
